File: backend/app/domain/rbac.py
"""Catálogo de permisos y navegación por rol (enforcement real vive en API)."""

from __future__ import annotations

from .entities import UserRole

PERMISSIONS: dict[UserRole, frozenset[str]] = {
    UserRole.OWNER: frozenset(
        {
            "user:manager",
            "tenant:settings",
            "project:staff",
            "evm.read",
            "alerts.read",
            "schedule:view",
            "cost:view",
            "cost:approve",
            "project:approve",
            "budget:approve",
        }
    ),
    UserRole.PMO: frozenset(
        {
            "project:manager",
            "schedule:edit",
            "schedule:view",
            "baseline.upload",
            "baseline.draft.edit",
            "baseline.freeze",
            "baseline.replan",
            "baseline.read",
            "evidence.approve",
            "evm.read",
            "alerts.read",
            "progress.cut.approve",
            "cost:approve",
            "budget:counter",
        }
    ),
    UserRole.ADMIN_OBRA: frozenset(
        {
            "baseline.read",
            "evm.read",
            "alerts.read",
            "labor.lock.read",
            "finance.retention.read",
        }
    ),
    UserRole.FINANCE: frozenset(
        {
            "baseline.read",
            "evm.read",
            "budget:propose",
            "cost.ac_upload",
            "labor.f30.upload",
            "labor.f30.approve",
            "finance.ep.process",
        }
    ),
    UserRole.COMMERCIAL: frozenset(
        {
            "evm.read",
            "alerts.read",
            "budget:approve",
            "cost:view",
            "schedule:view",
        }
    ),
    UserRole.OFICINA_TECNICA: frozenset(
        {
            "baseline.upload",
            "baseline.draft.edit",
            "baseline.read",
            "evidence.upload",
            "evidence.approve",
            "evm.read",
        }
    ),
    UserRole.WAREHOUSE: frozenset({"inventory:dispatch", "evm.read"}),
    UserRole.FIELD: frozenset(
        {
            "progress.submit",
            "evidence.upload",
            "cost.ac_upload",
            "baseline.read",
            "evm.read",
        }
    ),
    UserRole.SUBCONTRACTOR: frozenset({"evidence.upload", "inventory:request", "evm.read"}),
    UserRole.VIEWER: frozenset({"baseline.read", "evm.read", "alerts.read"}),
}

NAV_ITEMS: dict[UserRole, tuple[tuple[str, str], ...]] = {
    UserRole.OWNER: (
        ("/demo", "Dashboard"),
        ("/admin", "Administración"),
        ("/budget-approval", "Validar presupuestos"),
        ("/users", "Usuarios"),
        ("/profiles", "Perfiles"),
        ("/teams", "Equipos"),
        ("/areas", "Áreas y puestos"),
        ("/tenant", "Datos del tenant"),
    ),
    UserRole.PMO: (
        ("/demo", "Dashboard"),
        ("/projects", "Proyectos"),
        ("/gantt", "Carta Gantt"),
        ("/budget-review", "Validar presupuesto"),
        ("/cuts", "Cortes de control"),
        ("/evidence", "Evidencias"),
    ),
    UserRole.ADMIN_OBRA: (("/demo", "Dashboard"),),
    UserRole.FINANCE: (
        ("/demo", "Dashboard"),
        ("/budget", "Presupuesto de obras"),
        ("/finance", "Estados de pago"),
    ),
    UserRole.COMMERCIAL: (("/demo", "Dashboard"), ("/budget-approval", "Validar presupuestos")),
    UserRole.WAREHOUSE: (("/demo", "Dashboard"),),
    UserRole.OFICINA_TECNICA: (("/demo", "Dashboard"), ("/evidence", "Evidencias")),
    UserRole.FIELD: (
        ("/progress", "Registrar avance"),
        ("/costs", "Imputar AC"),
    ),
    UserRole.SUBCONTRACTOR: (("/progress", "Reportar avance"),),
    UserRole.VIEWER: (("/demo", "Dashboard"),),
}


def has_permission(role: UserRole, permission: str) -> bool:
    return permission in PERMISSIONS.get(role, frozenset())

File: apps/web/src/lib/rbac.ts
import type { UserRole } from "./types";

export const NAV_BY_ROLE: Record<UserRole, { href: string; label: string; group: string; mobile?: number }[]> = {
  owner: [
    { href: "/demo", label: "Inicio", group: "hoy", mobile: 1 },
    { href: "/admin", label: "Administración", group: "hoy" },
    { href: "/users", label: "Personas", group: "personas", mobile: 3 },
    { href: "/profiles", label: "Puestos", group: "personas" },
    { href: "/teams", label: "En la obra", group: "personas" },
    { href: "/areas", label: "Áreas", group: "personas" },
    { href: "/tenant", label: "Empresa", group: "empresa" },
    { href: "/budget-approval", label: "Aprobar $", group: "dinero" },
  ],
  pmo: [
    { href: "/demo", label: "Inicio", group: "hoy", mobile: 2 },
    { href: "/projects", label: "Obras", group: "obra", mobile: 5 },
    { href: "/gantt", label: "Calendario", group: "obra", mobile: 7 },
    { href: "/budget-review", label: "Revisar $", group: "dinero" },
    { href: "/cuts", label: "Cortes", group: "obra" },
    { href: "/evidence", label: "Evidencias", group: "obra" },
    { href: "/progress", label: "Avance", group: "obra", mobile: 4 },
  ],
  admin_obra: [
    { href: "/demo", label: "Inicio", group: "hoy", mobile: 2 },
    { href: "/progress", label: "Avance", group: "obra", mobile: 4 },
  ],
  finance: [
    { href: "/demo", label: "Inicio", group: "hoy", mobile: 2 },
    { href: "/budget", label: "Presupuesto", group: "dinero" },
    { href: "/finance", label: "Pagos", group: "dinero" },
    { href: "/progress", label: "Avance", group: "obra", mobile: 4 },
  ],
  commercial: [
    { href: "/demo", label: "Inicio", group: "hoy", mobile: 2 },
    { href: "/clients", label: "Clientes", group: "obra" },
    { href: "/budget-approval", label: "Aprobar $", group: "dinero" },
    { href: "/progress", label: "Avance", group: "obra", mobile: 4 },
  ],
  warehouse: [
    { href: "/demo", label: "Inicio", group: "hoy", mobile: 2 },
    { href: "/progress", label: "Avance", group: "obra", mobile: 4 },
  ],
  oficina_tecnica: [
    { href: "/demo", label: "Inicio", group: "hoy", mobile: 2 },
    { href: "/evidence", label: "Evidencias", group: "obra" },
    { href: "/progress", label: "Avance", group: "obra", mobile: 4 },
  ],
  field: [
    { href: "/demo", label: "Inicio", group: "hoy", mobile: 2 },
    { href: "/progress", label: "Avance", group: "obra", mobile: 4 },
    { href: "/costs", label: "Gastos", group: "dinero", mobile: 6 },
  ],
  subcontractor: [
    { href: "/demo", label: "Inicio", group: "hoy", mobile: 2 },
    { href: "/progress", label: "Avance", group: "obra", mobile: 4 },
  ],
  viewer: [{ href: "/demo", label: "Inicio", group: "hoy", mobile: 2 }],
};

export function navForRole(role: UserRole) {
  return NAV_BY_ROLE[role];
}
